File: py/structural/proxy.py
# Proxy controls access to another object, here delaying expensive creation.
# Modern Python note: proxies are useful for lazy loading, access checks, caching,
# remoting, and instrumentation when callers should keep the same interface.
class LargeImage:
    def __init__(self, path):
        print(f"loading {path}")
        self.path = path

    def display(self):
        return f"displaying {self.path}"


class ImageProxy:
    def __init__(self, path):
        self.path = path
        self._image = None

    def display(self):
        if self._image is None:
            self._image = LargeImage(self.path)
        return self._image.display()


if __name__ == "__main__":
    image = ImageProxy("photo.jpg")
    print("proxy created")
    print(image.display())

File: csharp/structural/proxy.cs
namespace Gof.ModernCSharp.Structural;

// Proxy controls access to another object, here delaying expensive creation.
// Modern C# note: useful for lazy loading, access checks, caching, remoting, and
// instrumentation when callers should keep the same interface.
public static class Proxy
{
    interface IImage { string Display(); }
    sealed class LargeImage(string path) : IImage
    {
        public string Display() => $"displaying {path}";
    }

    sealed class ImageProxy(string path) : IImage
    {
        readonly Lazy<LargeImage> _image = new(() => new LargeImage(path));
        public string Display() => _image.Value.Display();
    }

    public static string Run() => new ImageProxy("photo.jpg").Display();
}
